viscy/translation/evaluation.py

```diff
@@ -12,43 +15,254 @@
+    """Test runner for intensity metrics.
+
+    Parameters
+    ----------
+    metrics : list[str], optional
+        List of metrics to compute, by default ["mae", "mse", "ssim", "pearson"]
+    aggregate_epoch : bool, optional
+        Whether to aggregate results over the entire epoch, by default False
+    """
+
+    def __init__(
+        self,
+        metrics: list[str] = ["mae", "mse", "ssim", "pearson"],
+        aggregate_epoch: bool = False,
+    ) -> None:
+        super().__init__()
+        self.metrics = metrics
+        self.aggregate_epoch = aggregate_epoch
+        self._validate_metrics()
+
+    def _validate_metrics(self):
+        """Validate the metrics parameter."""
+        valid_metrics = ["mae", "mse", "ssim", "ms_ssim", "pearson", "cosine"]
+        for metric in self.metrics:
+            if metric not in valid_metrics:
+                raise ValueError(f"Metric '{metric}' not in {valid_metrics}")
+
+    def test_step(self, batch, batch_idx: int) -> None:
+        """Compute intensity metrics between prediction and target."""
+        from torchmetrics.functional import (
+            cosine_similarity,
+            mean_absolute_error,
+            mean_squared_error,
+            pearson_corrcoef,
+            structural_similarity_index_measure,
+        )
+
+        from viscy.translation.evaluation_metrics import ms_ssim_25d
+
+        pred = batch["pred"]
+        target = batch["target"]
+
+        # Dictionary to store computed metrics
+        metrics_dict = {
+            "position": batch["position_idx"][0] if "position_idx" in batch else -1,
+            "time": batch["time_idx"][0] if "time_idx" in batch else -1,
+        }
+
+        # Add metadata if available
+        if "cell_type" in batch:
+            metrics_dict["cell_type"] = batch["cell_type"]
+        if "organelle" in batch:
+            metrics_dict["organelle"] = batch["organelle"]
+        if "infection_condition" in batch:
+            metrics_dict["infection_condition"] = batch["infection_condition"]
+
+        # Compute metrics
+        for metric in self.metrics:
+            if metric == "mae":
+                metrics_dict["mae"] = mean_absolute_error(pred, target)
+            elif metric == "mse":
+                metrics_dict["mse"] = mean_squared_error(pred, target)
+            elif metric == "ssim":
+                # Handle different dimensionality cases
+                if pred.shape[0] > 1:  # 3D/2.5D case
+                    metrics_dict["ssim"] = structural_similarity_index_measure(
+                        (
+                            pred.squeeze(2)
+                            if pred.shape[2] == 1
+                            else pred[:, :, pred.shape[2] // 2]
+                        ),
+                        (
+                            target.squeeze(2)
+                            if target.shape[2] == 1
+                            else target[:, :, target.shape[2] // 2]
```

Comment:
Potentially dangerous: this could compute SSIM on a different region than the other metrics.